File: Practica 9/ifs.py
import random

s = "iCtSA"
a = ";eS"

d = s
maximo = 100
print("Tenemos que S - > " + s)
print("Tenemos que a - > " + a + "|" + str(chr(1013)))
print("Comenzamos con: ")
print(s)
cadenaFinal = ""
def generarCadena(cadena, tam, max):
	if tam == maximo:
		global cadenaFinal
		cadenaFinal = cadena
		return
	i = len(cadena)-1
	while(i>0):
		if cadena[i]=="A":
			print("Sustituyendo a A por ", end="")
			if random.randint(0,1):
				temp = cadena[:i] + "(" + a + ")" + cadena[i+1:]
				print(a)
				print(temp)
				generarCadena(temp, tam+1,max)
				return
			if(tam==0):
				temp = cadena[:i]
				print("Epsilon")
				print(temp+str(chr(1013)) + "->")
				print(temp)
				generarCadena(temp, tam+1, max+1)
				return
			temp = cadena[:i] + cadena[i+1:]
			tempEpsilon = cadena[:i] + str(chr(1013)) +cadena[i+1:]
			print("Epsilon")
			print(tempEpsilon + "->")
			print(temp)
			generarCadena(temp, tam+1, max+1)
			return
		if cadena[i]=="S":
			print("Sustituyendo a S")
			temp = cadena[:i] + "(" + s + ")" + cadena[i+1:]
			print(temp)
			generarCadena(temp, tam+1, max)
			return
		i-=1

generarCadena(s,0,0)
cadenaIfs = ""
tabulaciones = 0
for i in cadenaFinal:
	if i == "i":
		cadenaIfs+= "\t" * (tabulaciones) + "if"
	elif i == "C":
		cadenaIfs+="(C)"
	elif i == "t":
		tabulaciones+=1
		cadenaIfs+="{\n"
	elif i == ";":
		tabulaciones-=1
		cadenaIfs+="\n" + "\t" * (tabulaciones) +  "}"
	elif i == "e":
		cadenaIfs += "\n" + "\t" * (tabulaciones) +  "else{\n"
		tabulaciones+=1
	elif i == ")":
		tabulaciones-=1
		cadenaIfs += "\n" + "\t" * (tabulaciones) +  "}"
		if(tabulaciones == 0):
			break

fil = open("ifs.txt", "w")
fil.write(cadenaIfs)
fil.close()

File: Practica 9/ifs.cpp
#include<iostream>
#include <cstdlib>     /* srand, rand */
#include <ctime>
#include <fstream>

using namespace std;

string generarCadenaIfs(int numeroTab){
    string temp = "";
    for(int i = 0 ; i < numeroTab; i++){
        temp += "\t";
    }    
    return temp;
}

int main(){
    string s = "iCtSA";
    string a = ";eS";
    string cadenaFinal = s;
    std::ofstream outfile;
    outfile.open("ifs.txt", std::ios_base::out);
    double tam = 5;
    double i = tam - 1;
    double max = 100000;
    double count = 0;
    srand(time(0));
    int random;
    int parentesisSaltado = 0;
    while(count < max){
        if(cadenaFinal[i] == ')'){
            i--;
            continue;
        }
        if(cadenaFinal[i] == 'A'){
            random = rand() % 2;
            if (random) {
                cadenaFinal.replace(i, 1, "(;eS)");
                parentesisSaltado++;
                i+=4;
            }
            else{
                cadenaFinal.replace(i,1,"");
            }
        }
        if(cadenaFinal[i] == 'S'){
            cadenaFinal.replace(i,1,"(iCtSA)");
            parentesisSaltado++;
            i+=6;
        }
        i--;
        count++;
    }
    i+=parentesisSaltado;
    int tabulaciones = 0;
    string cadenaIfs = "";
    for(int j = 0; j < i; j++){
	    if (cadenaFinal[j] == 'i')
	    	cadenaIfs+= generarCadenaIfs(tabulaciones) + "if";
	    else if (cadenaFinal[j] == 'C')
	    	cadenaIfs+="(C)";
	    else if (cadenaFinal[j] == 't'){
	    	tabulaciones+=1;
	    	cadenaIfs+="{\n";
        }  
	    else if (cadenaFinal[j] == ';'){
	    	tabulaciones-=1;
	    	cadenaIfs+="\n" + generarCadenaIfs(tabulaciones) +  "}";
        }
	    else if (cadenaFinal[j] == 'e'){
	    	cadenaIfs += "\n" + generarCadenaIfs(tabulaciones) +  "else{\n";
	    	tabulaciones+=1;
        }
	    else if(cadenaFinal[j] == ')'){
	    	tabulaciones-=1;
	    	cadenaIfs += "\n" + generarCadenaIfs(tabulaciones) +  "}";
	    	if(tabulaciones == 0){
	    		break;
            }
        }
    }
    outfile << cadenaIfs <<  '\n';
}
